### PR TITLE
Fix APR's nill header issue; Compute APR based on last epoch

### PR DESCRIPTION
**Pull request by rlan35**
The issue why previously the header of last block of one epoch ago is always nil, is because the last epoch is actually set to current epoch whose last block is actually the current block to commit, thus it can not be found in blockchain.

Also hit an issue that when external node uses same BLS keys are harmony nodes, the logic is messed up, external nodes may get reward because harmony node signed the block.

Tested locally.

### CODE REVIEW

**Comment by stephen-tse** [comment]
looks good. let's inline these insights into the code?

"The issue why previously the header of last block of one epoch ago is always nil, is because the last epoch is actually set to current epoch whose last block is actually the current block to commit, thus it can not be found in blockchain.

Also hit an issue that when external node uses same BLS keys are harmony nodes, the logic is messed up, external nodes may get reward because harmony node signed the block.

File: consensus/votepower/roster.go [hunk 1]
```diff
@@ -204,7 +204,10 @@ func Compute(subComm *shard.Committee, epoch *big.Int) (*Roster, error) {
 			ourPercentage = ourPercentage.Add(member.OverallPercent)
 		}
 
-		roster.Voters[staked[i].BlsPublicKey] = &member
+		// TODO: make sure external user's BLS key can be same as harmony's bls keys
+		if _, ok := roster.Voters[staked[i].BlsPublicKey]; !ok {
```
Comment (LeoHChen) [inline consensus/votepower/roster.go]: do you need to print out a log  if the abnormal situation is found?

File: staking/apr/compute.go [hunk 2]
```diff
@@ -27,22 +29,21 @@ type Reader interface {
 }
 
 const (
-	secondsInYear = int64(31_557_600)
+	secondsInYear = int64(31557600)
```
Comment (LeoHChen) [inline staking/apr/compute.go]: why change this line?

File: staking/apr/compute.go [hunk 3]
```diff
@@ -55,152 +56,81 @@ func expectedRewardPerYear(
...
-
-	// result := numeric.NewDecFromBigInt(estimatedRewardPerYear).Quo(
-	// 	total,
-	// )
-	// return &result, nil
+	oneEpochAgo, zero :=
+		new(big.Int).Sub(block.Epoch(), common.Big1),
+		numeric.ZeroDec()
+
+	utils.Logger().Info().
```
Comment (LeoHChen) [inline staking/apr/compute.go]: debug?

File: staking/apr/compute.go [hunk 4]
```diff
@@ -55,152 +56,81 @@ func expectedRewardPerYear(
...
-	// total := numeric.NewDecFromBigInt(validatorNow.TotalDelegation())
-	// if total.IsZero() {
-	// 	return nil, errors.New("zero total delegation will cause div by zero")
-	// }
-
-	// result := numeric.NewDecFromBigInt(estimatedRewardPerYear).Quo(
-	// 	total,
-	// )
-	// return &result, nil
+	oneEpochAgo, zero :=
```
Comment (LeoHChen) [inline staking/apr/compute.go]: do you need to check if oneEpochAgo is greater than 0?

File: staking/apr/compute.go [hunk 5]
```diff
@@ -55,152 +56,81 @@ func expectedRewardPerYear(
...
+	)
+
+	if err != nil {
+		return &zero, nil
+	}
+
+	blockNumAtOneEpochAgo := shard.Schedule.EpochLastBlock(oneEpochAgo.Uint64())
+
+	headerOneEpochAgo := bc.GetHeaderByNumber(blockNumAtOneEpochAgo)
+	if block.Header() == nil || headerOneEpochAgo == nil || err != nil {
```
Comment (gupadhyaya) [inline staking/apr/compute.go]: `err != nil` already checked?

File: staking/apr/compute.go [hunk 6]
```diff
@@ -55,152 +56,81 @@ func expectedRewardPerYear(
...
+		Uint64("one-epoch-ago", oneEpochAgo.Uint64()).
+		Msg("apr - begin compute for validator ")
+
+	oneSnapshotAgo, err := bc.ReadValidatorSnapshotAtEpoch(
+		oneEpochAgo,
+		validatorNow.Address,
+	)
+
+	if err != nil {
+		return &zero, nil
```
Comment (gupadhyaya) [inline staking/apr/compute.go]: why discard the err?

File: staking/apr/compute.go [hunk 7]
```diff
@@ -55,152 +56,81 @@ func expectedRewardPerYear(
...
+		RawJSON("one-epoch-ago-header", []byte(headerOneEpochAgo.String())).
+		Msg("headers used for apr computation")
+
+	estimatedRewardPerYear, err := expectedRewardPerYear(
+		block.Header(), headerOneEpochAgo,
+		validatorNow, oneSnapshotAgo,
+	)
+
+	if err != nil {
+		return nil, err
```
Comment (gupadhyaya) [inline staking/apr/compute.go]: why return `nil` over `&zero` ? I am bit confused about the return value specification.

File: staking/apr/compute.go [hunk 8]
```diff
@@ -27,22 +29,21 @@ type Reader interface {
...
+	now, oneEpochAgo *block.Header,
+	curValidator, snapshotLastEpoch *staking.ValidatorWrapper,
 ) (*big.Int, error) {
-	oneTAgo, twoTAgo := oneEpochAgo.Time(), twoEpochAgo.Time()
+	timeNow, oneTAgo := now.Time(), oneEpochAgo.Time()
 	diffTime, diffReward :=
-		new(big.Int).Sub(twoTAgo, oneTAgo),
-		new(big.Int).Sub(twoSnapshotAgo.BlockReward, oneSnapshotAgo.BlockReward)
+		new(big.Int).Sub(timeNow, oneTAgo),
+		new(big.Int).Sub(curValidator.BlockReward, snapshotLastEpoch.BlockReward)
```
Comment (gupadhyaya) [inline staking/apr/compute.go]: do we need to take the absolute value here? as `diffReward` can be negative?

File: consensus/votepower/roster.go (hunk 1 above)
Comment (rlan35) [reply consensus/votepower/roster.go]: done.

File: staking/apr/compute.go (hunk 2 above)
Comment (rlan35) [reply staking/apr/compute.go]: it doesn't compile well in my IDE. Guess the style is not recognized universally

File: staking/apr/compute.go (hunk 3 above)
Comment (rlan35) [reply staking/apr/compute.go]: done

File: staking/apr/compute.go (hunk 4 above)
Comment (rlan35) [reply staking/apr/compute.go]: no, it should be and this only happens in staking epoch. And if not, the next part of the code to fetch snapshot will return nothing and it just returns.

File: staking/apr/compute.go (hunk 5 above)
Comment (rlan35) [reply staking/apr/compute.go]: good catch.

File: staking/apr/compute.go (hunk 6 above)
Comment (rlan35) [reply staking/apr/compute.go]: because this error should be a non-blocking error. We don't want to fail the block commitment because of this.

Comment (rlan35) [reply staking/apr/compute.go]: yea, we shouldn't. Fixed.

File: staking/apr/compute.go (hunk 7 above)
Comment (rlan35) [reply staking/apr/compute.go]: fixed. the errors should be returned.

File: staking/apr/compute.go (hunk 8 above)
Comment (rlan35) [reply staking/apr/compute.go]: diffReward can not. because reward is increasing only.

Comment (gupadhyaya) [reply staking/apr/compute.go]: true. I got confused with block reward.

**Pull request merged**